hack/tools/e2e-log-sync/main.go

```diff
@@ -0,0 +1,327 @@
+	logProwJobURL = flag.String("log-prow-job-url", "", "URL of the ProwJob to get the logs from")
+	lokiURL       = flag.String("loki-url", "http://localhost:3100/loki/api/v1/push", "Loki URL to push the logs to")
+)
+
+func main() {
+	flag.Parse()
+	if err := validateAndCompleteFlags(); err != nil {
+		fmt.Printf("Failed to validate flags: %v\n", err)
+		os.Exit(1)
+	}
+
+	if err := importLogs(); err != nil {
+		fmt.Printf("Failed to import logs: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func validateAndCompleteFlags() error {
+	if *logBucket == "" {
+		fmt.Printf("--log-bucket must not be empty\n")
+	}
+	if *logFolder == "" && *logProwJobURL == "" {
+		fmt.Printf("either --log-folder or --log-prow-job-url must not be empty\n")
+	}
+	if *logFolder != "" && *logProwJobURL != "" {
+		fmt.Printf("only one of --log-folder or --log-prow-job-url can be set\n")
+	}
+
+	if *logProwJobURL != "" {
+		var err error
+		*logBucket, *logFolder, err = calculateLogLocation(*logProwJobURL)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func calculateLogLocation(prowJobURL string) (string, string, error) {
+	u, err := url.Parse(prowJobURL)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to parse ProwJob URL %q", prowJobURL)
+	}
+
+	// Path has e.g. format:
+	// /view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/6189/pull-cluster-api-build-main/1496233282759561216
+	p := strings.TrimPrefix(u.Path, "/")
+	pathSegments := strings.Split(p, "/")
+
+	bucket := pathSegments[2]
+	folder := path.Join(pathSegments[3:]...)
+
+	return bucket, folder, nil
+}
+
+func importLogs() error {
+	ctx := context.Background()
+
+	// Create GCS client.
+	client, err := storage.NewClient(ctx, option.WithoutAuthentication())
+	if err != nil {
+		return fmt.Errorf("failed to create GCS client: %v", err)
+	}
+	defer client.Close()
+
+	// Get logs from GCS
+	logs, err := getLogsFromGCS(ctx, client, *logBucket, *logFolder)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get logs from GCS")
+	}
+
+	for logFile, ld := range logs {
+		// Prepare logs for Loki.
+		streams, err := prepareDataForLoki(ld)
+		if err != nil {
+			return errors.Wrapf(err, "failed to prepare data for Loki")
+		}
+
+		// Push logs to Loki.
+		if err := pushLogsToLoki(ctx, logFile, streams); err != nil {
+			return errors.Wrapf(err, "failed to push data to Loki")
+		}
+	}
+
+	return nil
+}
+
+// LogData represents the log file and metadata.
+type LogData struct {
+	logs     []byte
+	metadata []byte
+	fileName string
+}
+
+// getLogsFromGCS gets logs from GCS.
+func getLogsFromGCS(ctx context.Context, client *storage.Client, bucket, folder string) (map[string]LogData, error) {
+	klog.Infof("Getting logs from gs://%s/%s", bucket, folder)
+
+	// Create iterator to iterate over GCS files in folder.
+	query := &storage.Query{
+		Prefix:    folder,
+		Delimiter: "",
+	}
+	it := client.Bucket(bucket).Objects(ctx, query)
```

Comment:
Does it make sense to set a timeout on the context?

Reply:
Good point. I'm now only setting the context on this level (and I increased it to two minutes).

Pulling logs from a regular ProwJob takes ~ 2 sec or so.